### PR TITLE
feat: Modernize Breast Cancer Detection web app into an offline-first PWA (React + Tailwind + Zustand) (#1797)

### LINKED ISSUE #1797: Feature Request: Modernize Web App UI to a Responsive, Offline-First PWA (React + Tailwind + Zustand)
**Is your feature request related to a problem? Please describe.**
Currently, several of the machine learning web applications in this repository (e.g., Bird Species Classification, Breast Cancer Detection) rely on older HTML/Flask templates. While they successfully demonstrate the underlying models, the user interface presents a few challenges:
1. **Lack of Responsiveness:** The current templates are not fully optimized for mobile devices, which is increasingly how users interact with and test deployed models.
2. **Dated UX:** The interfaces feel slightly dated, which can detract from the impressive machine learning work happening in the backend.
3. **Network Dependency:** Users need a constant, stable internet connection to interact with the frontend, which limits accessibility.

**Describe the solution you'd like**
I propose a structural modernization of the frontend architecture for one of these web apps to greatly enhance usability and maintainability. My proposed stack is:
* **React & Vite:** To migrate the frontend to a fast, component-based Single Page Application (SPA), making future feature additions much easier.
* **Tailwind CSS:** To implement a clean, highly responsive, and modern user interface that works flawlessly across all device sizes.
* **Zustand:** For lightweight and efficient state management.
* **Progressive Web App (PWA) Integration:** By configuring the app as an offline-first PWA, users will be able to "install" the web app on their mobile devices and experience a much faster, app-like interface. This makes querying the ML models significantly more reliable.

**Describe alternatives you've considered**
I considered simply adding custom CSS or Bootstrap to the existing HTML files. However, this is just a cosmetic patch. It does not solve the lack of offline capabilities, nor does it provide the modular, scalable architecture that a modern framework like React offers for the long-term health of the repository.

**Additional context**
I am highly passionate about bridging the gap between complex ML models and seamless user experiences. I have direct experience building complex, AI-driven, offline-first PWAs using this exact stack (React, Vite, Zustand), so I am very comfortable executing this architectural shift quickly and cleanly. 

I would love to take ownership of this enhancement. **Could you please assign this issue to me?** I am ready to start working on a Pull Request immediately upon your approval!

### PR DESCRIPTION
**Pull request by Anijesh**
## Related Issue

Closes: #1797

- [x] Contributor (GSSoC)

---

### ✨ What does this PR do?

Modernizes the **Breast Cancer Detection** web app frontend into a responsive, installable,
**offline-first PWA** — React 19 + Vite 7 + Tailwind CSS 4 + Zustand 5 + Workbox — living at
`Breast Cancer Detection using DL with Webapp/Webapp/frontend/`.

**The original interface is not removed.** `/` still renders the existing Jinja template
exactly as before; the new SPA is served alongside it at `/app`.

### Why this app?

The issue named two candidates. **Bird Species Classification is no longer a Flask/HTML app** —
it was migrated to Streamlit, so the issue's premise no longer applies there. **Breast Cancer
Detection is still a genuine Flask + Jinja app**: `/` renders a template and `/predict` takes a
form POST and re-renders the whole page. That is exactly the situation the issue describes, so
it is the one modernized here.

### The three problems in the issue, and the fix for each

| Problem from the issue | Fix |
|---|---|
| **Not responsive** on mobile | Mobile-first Tailwind layout — one column on phones, two from `lg`, touch-sized controls, safe-area insets for notched devices |
| **Dated UX** — full page reload per prediction | SPA with async `fetch`, drag-and-drop dropzone with live preview, loading state, per-class confidence bars, persisted dark mode |
| **Needs a constant connection** | Installable PWA: Workbox pre-caches the app shell (12 entries, ~304 KB), results persist on-device, and scans made offline are queued and sent automatically on reconnect |

### 🧠 A deliberate design decision

`POST /api/predict` is registered **`NetworkOnly`** in the service worker. Caching a
prediction would mean potentially showing a stale *diagnosis* for a different scan — actively
harmful in a medical context. So instead of guessing, the app **queues** the request, tells the
user plainly why, and sends it the moment connectivity returns.

### 🔌 Backend changes — additive only

Three new routes in `webapp.py`; `/` and `/predict` are **untouched**:

* `GET /api/health` → `{"status":"ok","model_loaded":true,"classes":[...]}`
* `POST /api/predict` → JSON with predicted class, confidence and the full probability
  distribution (normalising the head if it is not already softmaxed)
* `GET /app` → serves the built SPA, with a helpful message if it has not been built yet

Errors return JSON with meaningful statuses: `400` bad upload, `503` model missing, `500`
inference failure.

---

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

---

## How Has This Been Tested?

**1. Unit tests — 31 passed** (Vitest, jsdom), covering the store state machine (selection,
loading, success, error, offline queue, history cap and eviction) and the pure presentation
helpers (confidence and byte formatting, relative time, file validation, probability ranking,
payload normalisation):

```text
✓ tests/format.test.js (15 tests)
✓ tests/store.test.js  (16 tests)

Test Files  2 passed (2)
     Tests  31 passed (31)
```

**2. Production build succeeds and emits real PWA artifacts:**

```text
dist/manifest.webmanifest         0.51 kB
dist/index.html                   1.06 kB │ gzip:  0.52 kB
dist/assets/index-*.css          27.82 kB │ gzip:  5.75 kB
dist/assets/index-*.js          214.73 kB │ gzip: 67.67 kB
✓ built in 686ms

PWA v1.3.0
mode      generateSW
precache  12 entries (304.06 KiB)
files generated
  dist/sw.js
  dist/workbox-33a84d7e.js
```

**3. The real Flask backend was run and exercised end-to-end** (not mocked). Because the
trained weights are not distributed with the repo, a throwaway model of the same input shape
was used to drive the genuine `cv2` + Keras path, then deleted:

```console
$ curl -s localhost:5000/api/health
{"classes":["Benign","Malignant","Normal"],"model_loaded":true,"status":"ok"}

$ curl -s -X POST -F "image=@scan.png" localhost:5000/api/predict
{"confidence":0.3556,"elapsed_ms":144.5,"model":"VGG16","prediction":"Normal",
 "probabilities":{"Benign":0.2929,"Malignant":0.3515,"Normal":0.3556}}   # sums to 1.0

$ curl -s -X POST localhost:5000/api/predict            # error path
{"error":"No image was uploaded."}

$ curl -s localhost:5000/app | head -c 60               # SPA is served
<!doctype html><html lang="en" class="h-full">...

$ curl -s localhost:5000/ | head -c 60                  # ORIGINAL route unchanged
<!DOCTYPE html><html lang="en">...
```

**4. The UI screenshots below are real renders** of the built app driven in Chrome via
Playwright — including the offline and reconnect flows, exercised by actually taking the
browser context offline.

---

## Screenshots

> Responses come from a stub returning a representative payload, since the trained weights are
> not in the repo. The renders themselves are real. The ultrasound shown is a **synthetic
> placeholder**, not patient data.

**Desktop — before analysis, and with a result**

| | |
|---|---|
| desktop empty | desktop result |

**Dark mode, and the offline queue** — note the app shell and past results stay fully usable
with no network, while the new scan is held rather than failing:

| | |
|---|---|
| dark | offline queued |

**Reconnected — the queued scan is sent automatically, no user action:**

auto retry

**Mobile (390×844) — empty, result, offline:**

| | | |
|---|---|---|
| mobile empty | mobile result | mobile offline |

---

## Checklist

- [x] My code follows the guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly wherever it was hard to understand.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my feature works.
- [ ] Any dependent changes have been merged and published in downstream modules. *(n/a — no dependent changes)*

---

### Notes for reviewers

* **`node_modules/` and `dist/` are not committed** — only source, config, `package-lock.json`,
  generated icons and screenshots.
* One thing worth knowing for future JS work in this repo: the root `.gitignore` line 32 is
  `lib/` (a Python virtualenv rule), which **silently excluded the app's `src/lib/` directory**.
  I caught it before committing and added a scoped `!src/lib/` negation in the frontend's own
  `.gitignore`, so the same trap does not bite the next contributor.
* No Python dependency was added — dev uses Vite's proxy, so there is no CORS setup, and
  production is served by Flask itself.
* The three routes added to `webapp.py` are purely additive; the original author's UI and
  README signature are untouched.
* Suggested labels: `type:feature`, `level:intermediate`, `gssoc:approved`.

### CODE REVIEW

**Comment by Anijesh** [comment]
@Sibam-Paul requesting review 🙏 — this closes #1797.

Since I am contributing from a fork I cannot apply labels myself. Could you please add `type:feature`, `level:intermediate` and `gssoc:approved`?

A few notes to save you time:
- **Bird Species Classification was not a valid target** — it has already been migrated to Streamlit, so the issue's "HTML/Flask templates" premise only holds for Breast Cancer Detection. That is the app modernized here.
- **The original UI still works.** `/` and `/predict` are untouched; the PWA is served alongside at `/app` via three additive routes.
- The screenshots in the description are **real Chrome renders** of the built app, including the offline-queue and auto-retry flows driven by actually taking the browser offline. The real Flask endpoints were also exercised with curl (output in the description).
- `node_modules` and `dist` are not committed.

One thing worth flagging for the repo generally: the root `.gitignore` line 32 is `lib/` (a Python virtualenv rule), which silently excluded this app's `src/lib/` folder — two essential source files would have been missing from the commit. I caught it and added a scoped negation in the frontend's own `.gitignore`, but any future JS project in this repo with a `lib/` directory will hit the same trap.

**Comment by Sibam-Paul** [comment]
bro come in discord @Anijesh need to discuss id specter102 pin me there

**Review by Sibam-Paul — APPROVED** [review]
lgtm

**Review by Sibam-Paul — APPROVED** [review]
